src/backend/distributed/planner/merge_planner.c

```diff
@@ -0,0 +1,523 @@
+/*-------------------------------------------------------------------------
+ *
+ * merge_planner.c
+ *
+ * This file contains functions to help plan MERGE queries.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include <stddef.h>
```

Comment:
(not directly related this part of the code)

One idea I had was to have a specific planning logic for `merge` command, similar to `insert .. select` (see `distributed_planner` -> `CreateInsertSelectPlan()`) as opposed to this PR, where we follow through UPDATE/DELETE code-paths.

In one hand, the current approach gives us some good leverage, we don't need to write code for many things. On the other hand, I'm slightly worried that some of the codes for `UPDATE`/`DELETE` that we rely on for `merge` won't be directly applicable to `merge`. So far, I have not seen such case, but wanted to note here explicitly so that you could also think about that.

Another future advantage of having a separate logic could be that we can have multiple ways of planning `merge` -- see potential repartition suggestion -- where we can more easily `retry` other ways of planning it.

At this point, I do not have a strong opinion on either way, but thought could be nice to document this idea for future

Reply:
Agree, there are pros and cons of both the approaches. Initially we started off with the notion that MERGE is a type of modification query, such as update/delete. But I believe as we progress and start including complex scenarios, such as reference tables, non-colocated tables and mix of reference tables with distributed tables, we should detangle this from UID queries.